[NPoS] Implements dynamic number of nominators

gpestana opened the pull request:
This PR refactors the implementation of the `ElectionDataProvider` bounds and implements a dynamic nomination quota per voter, based on the account's balance and an arbitrary curve. 

Instead of a fixed `MAX_NOMINATIONS` per nominator, the `ElectionDataProvider` uses an implementor of `trait NominationsQuota<Balance>`  to calculate the max votes per nominator. Since the number of votes per nominator is dynamic, the election data provider (pallet staking) keeps track of the size of the snapshot to ensure that the final snapshot size is bounded, even when the number of votes per voter is calculated on the fly. This PR implements an `ElectionSizeTracker` to achieve this goal. 

The `FixedNominationsQuota<const MAX: u32>` struct implements the nomination quota trait and can be used to set the maximum number of nominations per nominator to 16, which is the current behaviour.

**Lazy nomination quota checks**: Note that if a nominator's quota decreases _after_ the nomination has been done, all the nominations will be used. The nominations quota limits is only enacted _when an account performs a new nomination_, in which case, if the number of nomination exceeds the current nominator quota, the `fn nominate` extrinsic fails with `TooManyTargets`. 

Notable changes:

### Trait `NominationsQuota`
An implementor of `NominationsQuota` returns the maximum number of votes per nominator. `FixedNominationsQuota` implements an instance of a fixed number of nominations, similar to the current implementation when set to 16.

### `ElectionsBounds` and `DataProviderBounds`
The `MaxElectableTargets` and `TargetsBound` are replaced by `ElectionBounds`. `ElectionsBounds` defines the count and size limits of both voters and targets for an election. 

The `struct ElectionBounds` sets the upper bounds in size (MB) and number of voters/targets to request from the election data provider. The bounds are defined over two axis: number of elements and size (in MB) of the snapshot that is built from the election result.

### Struct `ElectionSizeTracker`
Keeps track of the amount of data (in size and number of elements) that an election data provider prepares in `get_npos_voters` and `get_npos_targets`. The election data provider uses an instance of `ElectionSizeTracker` to know when the number of voters exceeded the bounds in terms of size and return the election data to EPM.

### Changes to the `ElectionDataProvider` interface
Both `fn ElectionDataProvider::election_voters` and `fn ElectionDataProvider::electable_targets` take as an input an instance of `DataProviderBounds`, instead of an optional max in the of number of returned elements.

### New events
- `SnapshotVotersSizeExceeded`: emitted when the election snapshot was exhausted due to the voters size limit. 
- `SnapshotTargetsSizeExceeded`: emitted when the election snapshot was exhausted due to the targets size limit. 

### Runtime API
This PR also adds a new runtime API to allow clients to query the nominations quota for a given balance. 

---

polkadot companion: https://github.com/paritytech/polkadot/pull/6807

Related to and reviving this PR and discussion https://github.com/paritytech/substrate/pull/10340

Helps https://github.com/paritytech/substrate/issues/13069
Closes https://github.com/paritytech/substrate/issues/12968

gpestana committed:
Implements dynamic nominations per nominator

gpestana committed:
Adds SnapshotBounds and ElectionSizeTracker

gpestana committed:
Changes the ElectionDataProvider interface to receive ElectionBounds as input

gpestana committed:
Implements get_npos_voters with ElectionBounds

gpestana committed:
Implements get_npos_targets with ElectionBounds

gpestana committed:
Adds comments

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2173771

gpestana committed:
tests

kianenigma commented on frame/election-provider-multi-phase/src/lib.rs:
```diff
@@ -671,6 +671,14 @@ pub mod pallet {
 
+		// The limits of targets to include in the snapshot per block.
+		#[pallet::constant]
+		type TargetsBounds: Get<ElectionBounds>;
```
These should replace `MaxElectableTargets` et al

kianenigma commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -674,3 +673,100 @@ pub type BoundedSupportsOf<E> = BoundedSupports<
+/// The main purpose of this is to prevent mixing the order of similarly typed arguments (e.g. u32
+/// size and count).
+#[derive(Copy, Clone)]
+pub struct ElectionBoundsBuilder {
```
👍

kianenigma commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -674,3 +673,100 @@ pub type BoundedSupportsOf<E> = BoundedSupports<
+/// Ordering: when comparing two instances of `ElectionBounds`, the `count` has priority over the
+/// `size`, ie. if `A.count > B.count`, then `A > B` regardless of their relative `size`.
+#[derive(Clone, Copy, RuntimeDebug, scale_info::TypeInfo, Encode, Decode, Eq)]
+pub struct ElectionBounds {
+	/// The bound on number of elements. `None` means unbounded.
+	pub count: Option<u32>,
+	/// The bound on size, in bytes. `None` means unbounded.
+	pub size: Option<u32>,
+}
```
I think we can actually nest two instances of this:

```
struct DataProviderBound { cound, size } 
struct ElectionBounds { voters: DataProviderBound, targets: DataProviderBound } 
```

and have all functions take one argument. The builder patter can hopefully help keep the process of building common patters of this neat.

kianenigma commented on frame/election-provider-multi-phase/src/lib.rs:
```diff
@@ -1589,15 +1600,18 @@ impl<T: Config> Pallet<T> {
-							compute: ElectionCompute::Fallback,
-						})
+				T::Fallback::instant_elect(
+					ElectionBounds::new_unbounded(),
```
note that we don't overwrite a bound here but the fallback itself can have a bound. This is why we put unbounded here (`onchain::BoundedExecution`).

kianenigma reviewed (commented):
Looks in the right direction so far.

gpestana committed:
Truncates nomninations that exceed nominations quota; Old tests passing

gpestana committed:
Uses DataProviderBounds and ElectionBounds (to continue)

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2199278

gpestana commented:
The current logic truncates the nominations list of a voter based on the voter's nominations quota. We could also skip the voter if the target list > nomination quota. However, it may be hard for UIs/nominators to estimate the max number of nominations since its value is dynamic, and dropping the nomination may be too harsh, leading to annoying behaviour and may pose security issues if/when too many validators are dropped due to exceeding quota. 

```rust
targets.truncate(
	T::NominationsQuota::get_quota_safe(voter_weight.into()) as usize);
```

When the targets list is truncated, the event `NominationsQuotaExceeded {staker, exceeded_nominations}` is emitted, so that nominators/UIs can take action and adjust the nominations.

Open question: how will UIs deal with the fact that the nominations quota may change at any time? UIs need to check against the current state of the nominator to see if the nominator will be excluded or not, based on the nominations quota curve. We add a new extrinsic `Call::nominations_quota(origin, balance)` that emits an event `NominationsQuotaForBalance { balance, nominations_quota }` with the nomination quota for a given balance.

**Corollary**: Truncating the target list based on the nomination quota instead of dropping voters that exceed their quota means that there's no need to keep iterate over the voters list for at least `NPOS_MAX_ITERATIONS_COEFFICIENT * max_allowed_len`, as all the nominators will be added if they have at least 1 target nominated.

gpestana committed:
Finishes conversions - tests passing

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2202225

gpestana committed:
Refactor staking in babe mocks

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2202326

gpestana committed:
Replaces MaxElectableTargets and MaxElectingVoters with ElectionBounds; Adds more tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2203493

gpestana committed:
Merge branch 'master' into gpestana/staking-dynamic-nominators

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2203590

Ank4n commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -674,3 +674,176 @@ pub type BoundedSupportsOf<E> = BoundedSupports<
+
+#[derive(Clone, Copy, Eq, Default, Debug)]
+pub struct DataProviderBounds {
+	pub count: Option<u32>,
+	pub size: Option<u32>,
```
Perhaps give them a type alias as well?

gpestana committed:
Fixes nits; node compiling

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2205167

gpestana committed:
bechmarks

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2206985

gpestana committed:
removes nomination_quota extrinsic to request the nomination quota

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2217133

gpestana committed:
Lazy quota check, ie. at nominate time only

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2218992

gpestana committed:
remove non-working test (for now)

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2219039

gpestana committed:
tests lazy nominations quota when quota is lower than current number of nominated targets

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2223832

gpestana committed:
Adds runtime API and custom RPC call for clients to query the nominations quota for a given balance

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2224583

rossbulat reviewed (commented):
Nice implementations. The naming convention all make sense to me, the `get_quota` implementation is simple. If understood, the only way for apps to retrieve a quota is by calling the runtime call, this is all fine - it would be useful to be able to get the quota breakpoints so apps have a way to visualise the curve, e.g. this data: 

```
match balance.into() {
			// random quota per balance for testing
			0..=110 => MAX,
			111 => 0,
			222 => 2,
			333 => MAX + 10,
			_ => MAX,
}
```
Rather than having to hardcode it.

gpestana committed:
removes old rpc

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2459809

gpestana committed:
Merge branch 'master' into gpestana/staking-dynamic-nominators

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2459965

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2459971

gpestana committed:
Cosmetic touches

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2460605

gpestana committed:
All mocks working

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2460950

gpestana committed:
Fixes benchmarking mocks

ggwpez committed:
Fix V2 PoV benchmarking (#13485)

* Bump default 'additional_trie_layers' to two

The default here only works for extremely small runtimes, which have
no more than 16 storage prefices. This is changed to a "sane" default
of 2, which is save for runtimes with up to 4096 storage prefices (eg StorageValue).

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Update tests and test weights

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Fix PoV weights

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* ".git/.scripts/commands/bench/bench.sh" pallet dev pallet_balances

* ".git/.scripts/commands/bench/bench.sh" pallet dev pallet_message_queue

* ".git/.scripts/commands/bench/bench.sh" pallet dev pallet_glutton

* ".git/.scripts/commands/bench/bench.sh" pallet dev pallet_glutton

* Fix sanity check

>0 would also do as a check, but let's try this.

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

---------

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>
Co-authored-by: command-bot <>

gpestana commented on frame/election-provider-multi-phase/src/lib.rs:
```diff
@@ -1401,13 +1401,16 @@ impl<T: Config> Pallet<T> {
-		let voter_limit = T::MaxElectingVoters::get().saturated_into::<usize>();
+		let election_bounds = T::ElectionBounds::get();
+		let target_limit =
+			election_bounds.targets.count.unwrap_or(u32::MAX).saturated_into::<usize>();
```
This is a good point. I think this approach is OK iff the election bounds counts are set, in which case we could default to `0` too. This comes across as somewhat dangerous to me now. I see three options:
1. Fail election if the bounds are not defined;
2. Default to `0` (which probably result in a bad/failed election); or
3. Define sane defaults as consts (e.g. voters: 1000, targets: 1000) and use them as default.

I'm inclined towards option 3., but this may be a silent issue since the election will probably work as expected but the developers may want different bounds (although they are misconfigure). Any ideas?

gpestana replied:
OTOH, we have the size limit tracker putting an upper bound on unbounded (in terms of `count`). But the size bound may also be `None`, which basically means no size limit (in bytes).

This seems to be a configuration issue, so probably we should fail if no election bounds are defined or not set sane election bounds by default if not explicitly given.

gpestana replied:
IMO, the pallet developer should make sure the correct bounds are configured in the EPM pallet, otherwise we'll need to fail the election here, proceed with the election silently with 0 targets/voters or make sure the runtime has sensible bounds, which depends on the case (although this pallet is for internal use only).

davxy committed:
Move BEEFY code to consensus (#13484)

* Move beefy primitives to consensus dir
* Move beefy gadget to client consensus folder
* Rename beefy crates

yjhmelody committed:
chore: move genesis block builder to chain-spec crate. (#13427)

* chore: move genesis block builder to block builder crate.

* add missing file

* chore: move genesis block builder to sc-chain-spec

* Update client/chain-spec/src/genesis.rs

Co-authored-by: Bastian Köcher <git@kchr.de>

* Update test-utils/runtime/src/genesismap.rs

Co-authored-by: Bastian Köcher <git@kchr.de>

* Update test-utils/runtime/client/src/lib.rs

* fix warnings

* fix warnings

---------

Co-authored-by: Bastian Köcher <git@kchr.de>

gpestana commented on frame/staking/src/lib.rs:
```diff
@@ -761,6 +767,102 @@ impl<AccountId, Balance: HasCompact + Zero> UnappliedSlash<AccountId, Balance> {
+	}
+
+	// Returns the voter's nomination quota based on the quota implementation.
+	fn get_quota(balance: Balance) -> u32;
```
maybe we should collapse `get_quota` and `get_quota_safe` and make sure this call is always safe.

gpestana replied:
Maybe renaming `get_quota` -> `curve` and `get_quota_safe` -> `get_quota` would be the best option here.

gpestana committed:
nits

koute committed:
Speed up storage iteration from within the runtime (#13479)

* Speed up storage iteration from within the runtime

* Move the cached iterator into an `Option`

* Use `RefCell` in no_std

* Simplify the code slightly

* Use `Option::replace`

* Update doc comment for `next_storage_key_slow`

dmitry-markin committed:
Make unbounded channels size warning exact (part 1) (#13490)

* Replace `futures-channel` with `async-channel` in `out_events`

* Apply suggestions from code review

Co-authored-by: Koute <koute@users.noreply.github.com>

* Also print the backtrace of `send()` call

* Switch from `backtrace` crate to `std::backtrace`

* Remove outdated `backtrace` dependency

* Remove `backtrace` from `Cargo.lock`

---------

Co-authored-by: Koute <koute@users.noreply.github.com>

gpestana committed:
more tests

lazam committed:
Removal of Prometheus alerting rules deployment in cloud-infra (#13499)

gpestana committed:
renames trait methods

gpestana committed:
nit

gpestana commented:
bot fmt

Commit:
".git/.scripts/commands/fmt/fmt.sh"

andresilva committed:
sp-consensus: remove unused error variants (#13495)

pmikolajczyk41 committed:
Expose `ChargedAmount` (#13488)

* Expose `ChargedAmount`

* Fix imports

acatangiu committed:
sc-consensus-beefy: fix metrics: use correct names (#13494)


Signed-off-by: acatangiu <adrian@parity.io>

gpestana committed:
clippy fix

gpestana commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into gpestana/staking-dynamic-nominators

rossbulat commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into gpestana/staking-dynamic-nominators

rossbulat commented on frame/staking/src/lib.rs:
```diff
@@ -316,6 +317,7 @@ use sp_staking::{
 	EraIndex, SessionIndex,
 };
 use sp_std::{collections::btree_map::BTreeMap, prelude::*};
+
```
```suggestion
```

rossbulat commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -783,17 +784,47 @@ impl<T: Config> Pallet<T> {
+					if voters_size_tracker.try_register_voter(targets.len(), voter_bounds).is_err()
+					{
+						// no more space left for the election result, stop iterating.
+						Self::deposit_event(Event::<T>::SnapshotVotersSizeExceeded {
```
Provided the nominator does not  amend their exceeded-nominator list, is this event continuously fired every time a snapshot is created until they correct it? If so these events could pile up if we have lazy nominators & no UI (or permissionless call) to correct it.

rossbulat replied:
This comment was actually for a few lines up -  relating to the `NominationsQuotaExceeded` event.

rossbulat reviewed (commented):
This PR looks highly polished to me - is there any curve pre-determined that will slot into place when we are ready to migrate from `FixedNominationsQuota` to `NominationsQuota`? The only concern I annotated is with the potential repetition of the `NominationsQuotaExceeded` event, that could maybe undesirably populate a block with such events if we have many over-exceeded nominators. Apart from this I'd not hesitate to click Approve.

It is worth mentioning these changes briefly in high level pallet docs, either regarding the snapshot or the election algorithm stuff in the staking pallet?

gpestana commented:
> The only concern I annotated is with the potential repetition of the `SnapshotVotersSizeExceeded` event, that could maybe undesirably populate a block with such events if we have many over-exceeded nominators.

So you mean `NominationsQuotaExceeded`? This event is the one that may be emitted multiple times per election round (imagine if X voters had quota exceeded when doing the election round, the event will be emitted X times). The `SnapshotVotersSizeExceeded` may only happen once per election.

rossbulat commented:
Yes sorry, I meant `NominationsQuotaExceeded`.

gpestana committed:
removes NominationsQuotaExceeded event

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v0-9-39/2277/1

gpestana committed:
Update frame/staking/src/lib.rs

Co-authored-by: Ross Bulat <ross@parity.io>

gpestana committed:
adds back the npos_max_iter

gpestana committed:
Merge branch 'master' into gpestana/staking-dynamic-nominators

gpestana committed:
remove duplicate imports added after merge

gpestana commented:
bot fmt

gpestana commented:
bot cancel 70-eb05f2e6-1a96-4cf4-841f-3805c9c9031a

gpestana committed:
fmt

kianenigma commented on bin/node/runtime/src/lib.rs:
```diff
@@ -626,17 +630,15 @@ frame_election_provider_support::generate_solution_type!(
 );
 
 parameter_types! {
+	pub ElectionBoundsMultiPhase: ElectionBounds = ElectionBoundsBuilder::new().voters_count(40_000).targets_count(10_000).build();
```
Given that many ppl copy-pasta this code, let's use some smaller bounds, and add some comment about what they imply. This is a good resouce. This is a good resource: https://substrate.stackexchange.com/questions/2843/how-can-the-substrate-staking-pallet-be-used-in-a-parachain/3289#3289

kianenigma commented on frame/babe/src/mock.rs:
```diff
@@ -163,6 +166,7 @@ parameter_types! {
 	pub const SlashDeferDuration: EraIndex = 0;
 	pub const RewardCurve: &'static PiecewiseLinear<'static> = &REWARD_CURVE;
 	pub const OffendingValidatorsThreshold: Perbill = Perbill::from_percent(16);
+	pub static ElectionsBounds: ElectionBounds = ElectionBoundsBuilder::new().build();
```
This, and in probably many others places, need not be `static`.

kianenigma commented on frame/election-provider-multi-phase/src/lib.rs:
```diff
@@ -1412,13 +1413,16 @@ impl<T: Config> Pallet<T> {
+		let voters = T::DataProvider::electing_voters(election_bounds.voters)
 			.map_err(ElectionError::DataProvider)?;
 
 		if targets.len() > target_limit || voters.len() > voter_limit {
```
We can just get rid of this. It is double checking some bounds that are already "guaranteed" to be met. If we are to check, we should also check the byte size (if possible to do cheaply).

gpestana replied:
Added voter size checks here too.

kianenigma replied:
Your point about this pallet not dealing with byte size is correct, but in extension it would also not have to deal with `length` either. 

In principle, this pallet only knows that there is an `election_bounds`, which has `voters` and `targets`, and once passed to the `T::DataProvider`, they need to be `respected`.

If you were to do this pedantically, you would make the fields of `ElectionProviderBounds` private to hide this actually. 

I would formulate this then as:

```
let election_bounds = T::ElectionBounds::get();

let targets = T::DataProvider::electable_targets(election_bounds.targets)
	.and_then(|t| election_bounds.targets.ensure_respected(&v))
	.map_err(ElectionError::DataProvider)?;

let voters = T::DataProvider::electing_voters(election_bounds.voters)
	.and_then(|v| election_bounds.voters.ensure_respected(&v))
	.map_err(ElectionError::DataProvider)?;
```

and then inside `ensure_respected` implementation you can check length, and check byte-size again as well.

kianenigma commented on frame/election-provider-multi-phase/src/lib.rs:
```diff
@@ -1533,15 +1537,18 @@ impl<T: Config> Pallet<T> {
-							compute: ElectionCompute::Fallback,
-						})
+				T::Fallback::instant_elect(
+					DataProviderBounds::new_unbounded(),
```
Note that the fallback will have its own bounds configured by the runtime -- would be good to document this. 

Admittedly, this code looks strange, and we could improve this. 

What we want to express, is basically a `type ElectionBounds` and a `type FallbackBounds`.

gpestana replied:
What you are suggesting here is to add a new config associated type to EPM like

```rust
type FallbackElectionBounds = Get<ElectionBounds>; 
```
which will be set to unbounded in this pallet (since we don't have it setup atm)?

And then here we can call

```rust
let fallback_bounds = T::FallbackElectionBounds::get();
T::Fallback::instant_elect(fallback_bounds.voters, fallback_bounds.targets);
```

I think this looks cleaner and will open the possibility for us to add the fallback on-chain election just by changing configs later on (as we discussed it's probably wise to get it eventually).

kianenigma commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -662,16 +656,166 @@ impl<AccountId: IdentifierT, Accuracy: PerThing128, Balancing: Get<Option<Balanc
 sp_core::generate_feature_enabled_macro!(runtime_benchmarks_or_fuzz_enabled, any(feature = "runtime-benchmarks", feature = "fuzzing"), $);
+
+#[derive(Clone, Copy, Default, Debug)]
+pub struct DataProviderBounds {
```
undocumented public types 😱

kianenigma commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -662,16 +656,166 @@ impl<AccountId: IdentifierT, Accuracy: PerThing128, Balancing: Get<Option<Balanc
 	<E as ElectionProviderBase>::MaxWinners,
 >;
+/// Count bound of data provider bounds.
+pub type CountBound = u32;
```
If you want to make sure we are not mistaking the two, a type alias is hardly better, and a `newType` is actually what you want: `struct SizeBound(pub u32)` and such. Then it is impossible to mistake the two.

kianenigma commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -662,16 +656,166 @@ impl<AccountId: IdentifierT, Accuracy: PerThing128, Balancing: Get<Option<Balanc
+
+	/// Returns true if `given_size` or `given_count` exhausts `self.size` or `self_count`,
+	/// respectively.
+	pub fn exhausted(self, given_size: Option<SizeBound>, given_count: Option<CountBound>) -> bool {
```
For example, the call site could still easily mix `given_size` and `given_count` here, the type alias won't help a lot with that.

kianenigma commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -662,16 +656,166 @@ impl<AccountId: IdentifierT, Accuracy: PerThing128, Balancing: Get<Option<Balanc
+
+#[derive(Clone, Copy, Default, Debug)]
+pub struct DataProviderBounds {
+	pub count: Option<CountBound>,
```
FYI if you make the newtype, these can just remain raw `u32`, because they are linked to a named filed that clears out what they are. 

The main hassle is when you have a function, and two consecitive arguments are u32.

kianenigma commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -662,16 +656,166 @@ impl<AccountId: IdentifierT, Accuracy: PerThing128, Balancing: Get<Option<Balanc
+			targets: self.targets.unwrap_or_default(),
+		}
+	}
+}
```
all of this could go a new file, re-exported from root, and have very minimal tests for the builder (if any of it has significant logic to it)

kianenigma commented on frame/election-provider-support/src/onchain.rs:
```diff
@@ -85,13 +86,9 @@ pub trait Config {
-	type TargetsBound: Get<u32>;
+	/// Elections bounds, to use when calling into
+	/// [`Config::DataProvider`]. It might be overwritten in the `InstantElectionProvider` impl.
+	type ElectionBounds: Get<ElectionBounds>;
```
```suggestion
	type Bounds: Get<ElectionBounds>;
```

Because, well, we are in a crate who's name starts with election 😅

Or `DataProviderBounds` is more accurate, implying the this is how much it fetches from the data provider.

kianenigma commented on frame/election-provider-support/src/onchain.rs:
```diff
@@ -159,13 +155,14 @@ impl<T: Config> ElectionProviderBase for OnChainExecution<T> {
 	fn instant_elect(
-		forced_input_voters_bound: Option<u32>,
-		forced_input_target_bound: Option<u32>,
+		voters_bounds: DataProviderBounds,
```
I would keep the name `forced` here as it implies something pretty important

kianenigma commented on frame/election-provider-support/src/onchain.rs:
```diff
@@ -101,12 +98,11 @@ pub type OnChainBoundedSupportsOf<E> = BoundedSupports<
+	bounds: ElectionBounds,
 ) -> Result<OnChainBoundedSupportsOf<T>, Error> {
-	let voters = T::DataProvider::electing_voters(maybe_max_voters).map_err(Error::DataProvider)?;
+	let voters = T::DataProvider::electing_voters(bounds.voters).map_err(Error::DataProvider)?;
```
Here's a suggestion in `trait ElectionDataProvider`:

```
	fn fetch(
		bounds: ElectionBounds,
	) -> data_provider::Result<(Vec<VoterOf<Self>>, Vec<Self::AccountId>)> {
		Self::electing_voters(bounds.voters)
			.and_then(|voters| Ok((voters, Self::electable_targets(bounds.targets)?)))
	}
```

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -761,6 +766,102 @@ impl<AccountId, Balance: HasCompact + Zero> UnappliedSlash<AccountId, Balance> {
+/// A nomination quota that allows up to MAX nominations for all validators.
+pub struct FixedNominationsQuota<const MAX: u32>;
+impl<Balance, const MAX: u32> NominationsQuota<Balance> for FixedNominationsQuota<MAX> {
+	type MaxNominations = Self;
```
```suggestion
	type MaxNominations = ConstU32<MAX>;
```

should do and no need for the other impl

kianenigma commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -754,13 +755,15 @@ impl<T: Config> Pallet<T> {
 	///
 	/// This function is self-weighing as [`DispatchClass::Mandatory`].
-	pub fn get_npos_voters(maybe_max_len: Option<usize>) -> Vec<VoterOf<Self>> {
+	pub fn get_npos_voters(voter_bounds: DataProviderBounds) -> Vec<VoterOf<Self>> {
```
```suggestion
	pub fn get_npos_voters(bounds: DataProviderBounds) -> Vec<VoterOf<Self>> {
```

for similar reasons; things that are implied should not be repeated in names. The function name already implies `voter` in this case.

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -761,6 +766,102 @@ impl<AccountId, Balance: HasCompact + Zero> UnappliedSlash<AccountId, Balance> {
+	_marker: sp_std::marker::PhantomData<AccountId>,
+}
+
+impl<AccountId> ElectionSizeTracker<AccountId> {
```
would like to see independent tests for this, I recall original PR had it.

gpestana replied:
There are unit tests for the `ElectionSizeTracker` here https://github.com/paritytech/substrate/pull/12970/files#diff-64c049b4e94a55bdeaf757c725cf7d1df623b754557af4ced157024a943c4703R5908 and the boundaries are also implicitly tested across other unit tests. I'll see if the test coverage can be improved though.

gpestana replied:
Check the previous tests, add more coverage and check the size_hint option.

kianenigma commented on frame/staking/src/pallet/mod.rs:
```diff
@@ -272,6 +271,15 @@ pub mod pallet {
+	/// Maximum limit of nominations per nominator, regardless of `T::NominationsQuota`.
+	#[pallet::extra_constants]
+	impl<T: Config> Pallet<T> {
+		#[pallet::constant_name(MaxNominations)]
```
Wasn't there supposed to be a runtime API for this as well? would be good to document in code and in general how this should be used next to the runtime API

gpestana replied:
The idea here was to expose the `AbsoluteMaxNominations` configured, but I think this is not that useful anymore. Removing it.

kianenigma reviewed (changes requested):
Mostly good, but need some further fixes.

gpestana committed:
Adds comment in public struct; Refactors CountBound and SizeCount to struct

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2560101

gpestana committed:
addresses various pr comments

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2560639

gpestana committed:
PR comment reviews

gpestana committed:
Fixes on-chain election bounds and related code

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2586734

gpestana committed:
EPM checks the size of the voter list returned by the data provider

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2587682

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2587667

gpestana committed:
cosmetic changes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2587778

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2587764

gpestana committed:
updates e2e tests mock

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2587879

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2587864

gpestana committed:
Adds more tests for size tracker and refactors code

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2589701

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2589702

gpestana committed:
Adds back only_iterates_max_2_times_max_allowed_len test

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2589752

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2589751

gpestana committed:
Refactor

gpestana committed:
removes unecessary dependency

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2589875

gpestana committed:
Merge branch 'master' into gpestana/staking-dynamic-nominators

gpestana commented:
bot bench $ pallet dev pallet-election-provider-multi-phase

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2590002

gpestana commented:
bot bench $ pallet dev pallet-election-provider-multi-phase

gpestana committed:
empty commit -- restart all stuck CI jobs

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2592424

gpestana committed:
restarts ci jobs

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2624922

gpestana committed:
Renames ElectionBounds -> Bounds in benchmarking mocks et al

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2630332

gpestana committed:
updates mocks

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2630485

gpestana committed:
Merge branch 'master' into gpestana/staking-dynamic-nominators

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -761,6 +766,105 @@ impl<AccountId, Balance: HasCompact + Zero> UnappliedSlash<AccountId, Balance> {
+
+/// A static tracker for the election data snapshot.
+///
+/// Computes the (SCALE) encoded byte length of a snapshot based on static rules, without any actual
```
```suggestion
impl<Balance, T: Get<u32>> NominationsQuota<Balance> for T {
	type MaxNominations: T;
	fn curve(_) -> u32 { T::get() }
}
```

Mostly sharing it because it is a nice trait gymnastics, but it is probably not as readable and clear `FixedNominationsQuota` having an explicit name. Up to you. We could also have both.

kianenigma commented on Cargo.lock:
```diff
@@ -1689,7 +1694,7 @@ checksum = "4fb810d30a7c1953f91334de7244731fc3f3c10d7fe163338a35b9f640960321"
  "proc-macro2",
  "quote",
- "syn",
+ "syn 1.0.109",
```
a rebase should remove most of these? not sure why it is happening.

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -761,6 +766,105 @@ impl<AccountId, Balance: HasCompact + Zero> UnappliedSlash<AccountId, Balance> {
 	}
 }
 
+/// Something that defines the maximum number of nominations per nominator based on a curve.
```
In a follow-up, you can also re-pot these types and traits into a new file, or here.

kianenigma commented on frame/staking/src/tests.rs:
```diff
@@ -5775,6 +5960,113 @@ fn set_min_commission_works_with_admin_origin() {
 	})
 }
 
+mod election_size_tracker {
```
Yeah, I would def. move this size tracker and its tests into a new file.

kianenigma commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -662,16 +658,235 @@ impl<AccountId: IdentifierT, Accuracy: PerThing128, Balancing: Get<Option<Balanc
+
+	/// Returns an instance of `Self` that is constructed by capping both the `count` and `size`
+	/// fields. If `self` is None, overwrite it with the provided bounds.
+	pub fn max(self, bounds: DataProviderBounds) -> Self {
```
If you implement `std::cmp` for `DataProviderBounds` this will be given by default.

A small test set of tests for this would also be good.

gpestana replied:
The problem I see with implementing `std::cmp` over `DataProviderBounds` is that we can't tell which is the right order given the data provider bounds are defined over 2 axis: voters and targets. E.g. if `a.voters` > `b.voters` and `a.targets` < `b.targets`, how should we return the ordering? we could set e.g. voters to have precedence over targets, but then I'm not sure `max` would work as expected.

Added more tests.

kianenigma commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -662,16 +658,235 @@ impl<AccountId: IdentifierT, Accuracy: PerThing128, Balancing: Get<Option<Balanc
+	}
+}
+
+/// The limits of an election result. The bounds are defined over the count of element of the
```
```suggestion
/// The limits of an election snapshot size. The bounds are defined over the count of element of the
```

kianenigma commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -662,16 +658,235 @@ impl<AccountId: IdentifierT, Accuracy: PerThing128, Balancing: Get<Option<Balanc
+		ElectionBoundsBuilder { voters: Some(bounds.voters), targets: Some(bounds.targets) }
+	}
+
+	// Sets the voters count bounds.
```
```suggestion
	/// Sets the voters count bounds.
```

kianenigma commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -662,16 +658,235 @@ impl<AccountId: IdentifierT, Accuracy: PerThing128, Balancing: Get<Option<Balanc
+	}
+
+	/// Returns a new election bounds builder from an instance of `ElectionBounds`.
+	pub fn from(bounds: ElectionBounds) -> Self {
+		ElectionBoundsBuilder { voters: Some(bounds.voters), targets: Some(bounds.targets) }
+	}
```
why not a proper `From<ElectionBounds>`?

kianenigma commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -662,16 +658,235 @@ impl<AccountId: IdentifierT, Accuracy: PerThing128, Balancing: Get<Option<Balanc
+
+	// Sets the voters size bounds.
+	pub fn voters_size(mut self, size: SizeBound) -> Self {
+		self.voters = self.voters.map_or(
```
```suggestion
	/// Sets the voters size bounds.
```

kianenigma replied:
a few other instances in this file.

kianenigma commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -662,16 +658,235 @@ impl<AccountId: IdentifierT, Accuracy: PerThing128, Balancing: Get<Option<Balanc
+	pub targets: DataProviderBounds,
+}
+
+/// Utility builder for [`ElectionBounds`].
```
```suggestion
/// Utility builder for [`ElectionBounds`].
///
/// ## Examples
///
/// ```
/// use frame_election_provider_support::ElectionBoundsBuilder;
///
/// # fn main() {
///
/// # }
/// ```
```
In the spirit of an up the docs game? I think compiling test in rust-doc is always a great idea. You can cover a few example of how the builder can be used to build common example, or edge-cases, if any.

kianenigma replied:
Basically I would move some or all of `mod elections_bounds` into rust-docs.

kianenigma commented on frame/election-provider-support/src/onchain.rs:
```diff
@@ -159,13 +156,15 @@ impl<T: Config> ElectionProviderBase for OnChainExecution<T> {
-			Some(T::TargetsBound::get().min(forced_input_target_bound.unwrap_or(u32::MAX)) as usize),
-		)
+		let elections_bounds = ElectionBoundsBuilder::from(T::Bounds::get())
+			.voters_or_lower(forced_input_voters_bounds)
```
I think similarly, if you implement `cmp` for `DataProviderBounds`, you can do: 

```
let elections_bounds = ElectionBoundsBuilder::from(T::Bounds::get().min(forced_bounds))
```

which also implies the input of the function could be simplified into one `ElectionBounds` rather than two `DataProviderBounds`.

kianenigma replied:
The API is overall still very confusing :( let's make a ticket to improve?

kianenigma commented on frame/election-provider-multi-phase/src/lib.rs:
```diff
@@ -1412,18 +1415,32 @@ impl<T: Config> Pallet<T> {
+		let election_bounds = T::ElectionBounds::get();
+		let target_limit = election_bounds
+			.targets
+			.count
```
```suggestion
			.count.map_or(usize::MAX, |c| c.0.saturated_into::<usize>())
```

ruseinov replied:
@kianenigma do you think `usize::MAX` is a good idea? I'm kinda afraid it's going to be too large.

kianenigma commented on frame/election-provider-multi-phase/src/signed.rs:
```diff
@@ -561,7 +561,8 @@ mod tests {
 		ExtBuilder::default().build_and_execute(|| {
 			// given a reduced expectation of maximum electable targets
-			MaxElectableTargets::set(2);
+			let new_bounds = crate::ElectionBoundsBuilder::new().targets_count(2.into()).build();
```
Can probably be moved to a confguration function on `ExtBuilder`.

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -761,6 +766,105 @@ impl<AccountId, Balance: HasCompact + Zero> UnappliedSlash<AccountId, Balance> {
 
+/// Something that defines the maximum number of nominations per nominator based on a curve.
+pub trait NominationsQuota<Balance> {
+	/// Maximum number of nominations. The method `curve` implements the nomination quota curve and
```
This doc seems to describe the whole trait more so than this type item.

This type items should be documented as "what is typically used with BoundedVec" etc etc.

kianenigma commented on frame/staking/src/pallet/mod.rs:
```diff
@@ -1161,7 +1163,11 @@ pub mod pallet {
-			ensure!(targets.len() <= T::MaxNominations::get() as usize, Error::<T>::TooManyTargets);
+			ensure!(
+				targets.len() <=
+					T::NominationsQuota::get_quota(Self::weight_of(&stash).into()) as usize,
```
I don't see any reason to use `weight_of` here -- `get_quota` needs a balance, and we aleady have it at hand, namely via `ledger.active` (`weight_of` is reading this internally, can double check).

ruseinov replied:
`weight_of` is `VoteWeight`, so it's using `ledger.active`, but then `CurrencyToVote`'ing it.

gpestana replied:
Changed to using `ledger.active` directly.

kianenigma commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -783,17 +786,37 @@ impl<T: Config> Pallet<T> {
 			} else if Validators::<T>::contains_key(&voter) {
 				// if this voter is a validator:
+				if voters_size_tracker.try_register_voter(1, bounds).is_err() {
+					// no more space left for the election result, stop iterating over.
```
```suggestion
					// no more space left for the election snapshot, stop iterating over.
```

kianenigma commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -807,8 +830,8 @@ impl<T: Config> Pallet<T> {
-				// `T::MaxNominations`. The latter can rarely happen, and is not really an emergency
-				// or bug if it does.
+				// `T::NominationsQuota::get_quota`. The latter can rarely happen, and is not
+				// really an emergency or bug if it does.
```
Should be `defensive!` instead of log warn.

gpestana replied:
Done.

kianenigma commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -818,7 +841,7 @@ impl<T: Config> Pallet<T> {
 
 		// all_voters should have not re-allocated.
-		debug_assert!(all_voters.capacity() == max_allowed_len);
+		debug_assert!(all_voters.capacity() == max_allowed_len as usize);
```
renaming `max_allowed_len` to `final_predicted_len` might help.

kianenigma commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -841,13 +864,21 @@ impl<T: Config> Pallet<T> {
+			let all_target_count = T::TargetList::count();
+			target_bounds
+				.count
+				.unwrap_or(all_target_count.into())
```
I think this chain of calls is a bit hard to follow, but not sure how to improve it either.

kianenigma commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -976,43 +1007,42 @@ impl<T: Config> Pallet<T> {
-		debug_assert!(maybe_max_len.map_or(true, |max| voters.len() <= max));
+		let voters = Self::get_npos_voters(bounds);
+
+		debug_assert!(!bounds.exhausted(
```
This is basically being checked in EPM, and something that we have discussed there as well. Fine with checking it here as well.

kianenigma commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -976,43 +1007,42 @@ impl<T: Config> Pallet<T> {
+		let voters = Self::get_npos_voters(bounds);
+
+		debug_assert!(!bounds.exhausted(
+			SizeBound(voters.encoded_size().saturating_sub(1) as u32).into(),
```
Why `.saturating_sub(1)`?

gpestana replied:
It doesn't really need it, removing it.

kianenigma commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -976,43 +1007,42 @@ impl<T: Config> Pallet<T> {
+	fn electable_targets(bounds: DataProviderBounds) -> data_provider::Result<Vec<T::AccountId>> {
 		let target_count = T::TargetList::count();
 
 		// We can't handle this case yet -- return an error.
```
Would be good to link to Roman's issue about TargetList here. That's the fix.

kianenigma commented on frame/staking/src/pallet/impls.rs:
```diff
@@ -976,43 +1007,42 @@ impl<T: Config> Pallet<T> {
 
 		// We can't handle this case yet -- return an error.
-		if maybe_max_len.map_or(false, |max_len| target_count > max_len as u32) {
+		if bounds.exhausted(None, CountBound(target_count as u32).into()) {
```
Why not check the size here as well? The `encoded_size` of `target_count` targets is `compact_u32_prefix(target_count) + target_count * size_of(AccountId)`.

gpestana replied:
Done, we're checking the target's size bound now.

kianenigma commented on frame/staking/src/lib.rs:
```diff
@@ -761,6 +766,105 @@ impl<AccountId, Balance: HasCompact + Zero> UnappliedSlash<AccountId, Balance> {
+	pub fn encoded_size(votes: usize) -> usize {
+		Self::length_prefix(votes)
+			.saturating_add(votes * sp_std::mem::size_of::<AccountId>())
+			.saturating_add(sp_std::mem::size_of::<frame_election_provider_support::VoteWeight>())
```
I am doubting my original decision here to use `mem::size_of`. We are relying on the fact that `AccountId` is something that is only stored in the stack, and therefore: `sp_std::mem::size_of::<AccountId>() == some_account_id.encoded_size()`. In reality, i think we need to use `encoded_size` here. 

I suggest two things: 

* Play around with `encoded_size` and `size_of` for more insight into this. 
* We should double check the size bounding extensively before using it in production. Perhaps we can mark the API somehow as `unsafe` or similiar to make sure it is being used with caution. 

Recall that the end goal of this PR is to make sure the code path in EPM does not run out of allocation memory. Some of the PRs that fixed the OOM issue at https://polkadot.network/blog/a-polkadot-postmortem-24-05-2021/. 

I am pretty sure that with extnsive testing, especially with this PR, we can grow the size of nominator snapshot safely to something like `count: 40,000 | size: 14MB`, but we need to do really really extensive testing on it e2e before risking anything for it.

kianenigma replied:
All of this can be follow-ups. For now I only suggest making sure that there are some warnings in place when using `size` bound, either via compiler warning, or in the docs.

gpestana replied:
I did some testing around the usage of `sp_std::mem::size_of`, [`Encoded::size_hint`](https://docs.rs/parity-scale-codec/latest/parity_scale_codec/trait.Encode.html#method.size_hint) and `encoded_size`. I think that the implementation is safer with `size_hint` and it achieves our goals: 1) no need to do extra allocations/encoding while iterating through the voters and 2) the resulting size of the tracker matches with the final `encoded_size` of the voters.

I also refactored all the code and tests related to the static size tracker to its own module and added more docs.

I'm now thinking about a way to add `try_state` checks that would allow us to extensively tests the size bounds and the static tracker. Once we have those implemented, we can run `follow-chain` using a runtime with different bounds and `try-state` enabled, so we can verify if the checks and configs work as expected.

kianenigma commented on frame/staking/src/tests.rs:
```diff
@@ -4618,6 +4597,173 @@ mod election_data_provider {
+			// set size bounds that allows only for 1 voter.
+			let bounds = ElectionBoundsBuilder::new().voters_size(26.into()).build();
+			let elected = Staking::electing_voters(bounds.voters).unwrap();
+			assert!(elected.encoded_size() <= 26 as usize);
```
```suggestion
			assert!(elected.encoded_size() == 26 as usize);
```

Try this out: If the account Id type is something that is allocated on the heap (`Vec<u8>`, or even `Box<u128>`), I think our existing code is wrong becuase it uses `size_of` instead of `encoded_size`.

kianenigma commented on frame/staking/src/tests.rs:
```diff
@@ -4618,6 +4597,173 @@ mod election_data_provider {
+	fn nomination_quota_checks_at_nominate_works() {
+		ExtBuilder::default().nominate(false).build_and_execute(|| {
+			// stash bond of 222 has a nomination quota of 2 targets.
+			bond(61, 60, 222);
```
```suggestion
			bond(61, 60, 222);
			assert_eq(Staking::api_nomination_quota(222), 2);
```

kianenigma commented on frame/staking/src/tests.rs:
```diff
@@ -4618,6 +4597,173 @@ mod election_data_provider {
+			.build_and_execute(|| {
+				// unbond 78 from stash 60 so that it's bonded balance is 222, which has a lower
+				// nomination quota than at nomination time (max 2 targets).
+				assert_ok!(Staking::unbond(RuntimeOrigin::signed(60), 78));
```
Similarly, would be good to assert on `api_nomiantion_quota` of `60` here.

kianenigma reviewed (approved):
I've suggested a lot of potential further improvements, some for now, and some as follow-up. But, assuming they are all treated reasonably, this is good to go.

Audit can certainly start. 

This is a good opportunity to use `try-runtime fast-forward` to make sure the eg Polkadot's next election snapshot is created correctly with the new logic here.

ruseinov commented on bin/node/runtime/src/lib.rs:
```diff
@@ -640,17 +644,19 @@ frame_election_provider_support::generate_solution_type!(
 		Accuracy = sp_runtime::PerU16,
-		MaxVoters = MaxElectingVoters,
+		MaxVoters = MaxNominations,
 	>(16)
```
Curious if this `16` could now be replaced with a const `MAX_QUOTA_NOMINATIONS`.  Might not be possible due to macro limitations.

ruseinov commented on frame/election-provider-multi-phase/src/lib.rs:
```diff
@@ -1089,13 +1086,19 @@ pub mod pallet {
-					},
-				)?;
+			let election_bounds = ElectionBoundsBuilder::new()
+				.voters_count(maybe_max_voters.unwrap_or(u32::MAX).into())
```
I'm not 100% sure `u32::MAX` is a safe choice here. Is it not going to blow up?

ruseinov commented on frame/election-provider-multi-phase/src/lib.rs:
```diff
@@ -1089,13 +1086,19 @@ pub mod pallet {
+			let election_bounds = ElectionBoundsBuilder::new()
+				.voters_count(maybe_max_voters.unwrap_or(u32::MAX).into())
+				.targets_count(maybe_max_targets.unwrap_or(u32::MAX).into())
+				.build();
```
Same as above.

ruseinov commented on frame/election-provider-multi-phase/src/lib.rs:
```diff
@@ -1412,18 +1415,32 @@ impl<T: Config> Pallet<T> {
+		let voter_limit = election_bounds
+			.voters
+			.count
+			.unwrap_or(u32::MAX.into())
```
Same here.

ruseinov commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -662,16 +658,235 @@ impl<AccountId: IdentifierT, Accuracy: PerThing128, Balancing: Get<Option<Balanc
+impl Add for CountBound {
+	type Output = Self;
+	fn add(self, rhs: Self) -> Self::Output {
+		CountBound(self.0 + rhs.0)
```
what happens in case of overflow?
I've seen some `u32::MAX` before, so I guess it's not impossible.

gpestana replied:
Refactored to use `saturating_add` here.

ruseinov commented on frame/election-provider-support/src/lib.rs:
```diff
@@ -662,16 +658,235 @@ impl<AccountId: IdentifierT, Accuracy: PerThing128, Balancing: Get<Option<Balanc
+impl Add for SizeBound {
+	type Output = Self;
+	fn add(self, rhs: Self) -> Self::Output {
+		SizeBound(self.0 + rhs.0)
```
nit: Same here, curious if we need to cater for overflow here to be precise.

ruseinov reviewed (approved):
Overall seems alright to me, not so sure about `u32::MAX` usages and `Add` implementation that does not cater for overflows.

gpestana committed:
Update frame/election-provider-support/src/lib.rs

Co-authored-by: Kian Paimani <5588131+kianenigma@users.noreply.github.com>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2664151

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2664152

gpestana committed:
Update frame/staking/src/pallet/impls.rs

Co-authored-by: Kian Paimani <5588131+kianenigma@users.noreply.github.com>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2664203

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2664202

gpestana committed:
Merge branch 'master' into gpestana/staking-dynamic-nominators

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2664406

gpestana committed:
Update frame/election-provider-support/src/lib.rs

Co-authored-by: Kian Paimani <5588131+kianenigma@users.noreply.github.com>

gpestana committed:
Update frame/staking/src/tests.rs

Co-authored-by: Kian Paimani <5588131+kianenigma@users.noreply.github.com>

gpestana committed:
more checks in api_nominations_quota in tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2666562

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2666563

gpestana committed:
Improves docs

gpestana committed:
fixes e2e tests

gpestana committed:
Uses size_hint rather than mem::size_of in size tracker; Refactor size tracker to own module

gpestana committed:
Merge branch 'master' into gpestana/staking-dynamic-nominators

gpestana committed:
nits from reviews

gpestana committed:
Merge branch 'master' into gpestana/staking-dynamic-nominators

gpestana committed:
Refactors bounds to own module; improves docs

gpestana committed:
More tests and docs

gpestana committed:
fixes docs

gpestana committed:
Fixes benchmarks

gpestana committed:
Fixes rust docs

gpestana committed:
fixes bags-list remote-ext-tests

gpestana committed:
Simplify bound checks in create_snapshot_external

gpestana committed:
Merge branch 'master' into gpestana/staking-dynamic-nominators

gpestana commented:
bot rebase

gpestana commented:
bot rebase

gpestana committed:
Merge branch 'master' into gpestana/staking-dynamic-nominators

gpestana committed:
Adds target size check in get_npos_targets

FatemeShirazi commented:
We will look into empirical data and suggest a curve by about next week.

gpestana committed:
Merge branch 'master' into gpestana/staking-dynamic-nominators

gpestana commented:
bot fmt

Commit:
".git/.scripts/commands/fmt/fmt.sh"

gpestana committed:
restart ci

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3336846

gpestana committed:
rust doc fixes and cosmetic nits

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3337184

gpestana commented:
bot rebase

gpestana commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into gpestana/staking-dynamic-nominators

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3346196

gpestana commented:
bot fmt

gpestana commented:
bot clean

gpestana commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into gpestana/staking-dynamic-nominators

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3352741

gpestana committed:
rollback upgrade on parity-scale-codec version (unecessary)

gpestana committed:
reset cargo lock, no need to update it

gpestana committed:
Merge branch 'master' into gpestana/staking-dynamic-nominators

gpestana commented:
bot merge

arkpar commented on frame/staking/src/tests.rs:
```diff
@@ -5120,50 +5297,55 @@ fn min_commission_works() {
 
 #[test]
-fn change_of_max_nominations() {
+#[should_panic]
```
Is this intended? If so, Is test supposed to panic at the last assertion?
